README.md
Docker-based integration tests
=====
[![PyPI version](https://img.shields.io/pypi/v/pytest-docker?color=green)](https://pypi.org/project/pytest-docker/)
[![Python versions](https://img.shields.io/pypi/pyversions/pytest-docker)](https://pypi.org/project/pytest-docker/)
[![Code style](https://img.shields.io/badge/formatted%20with-black-black)](https://github.com/psf/black)


# Description
Simple [pytest](http://doc.pytest.org/) fixtures that help you write integration
tests with Docker and [docker-compose](https://docs.docker.com/compose/).
Specify all necessary containers in a `docker-compose.yml` file and and
`pytest-docker` will spin them up for the duration of your tests.

This package is tested with Python versions `3.5`, `3.6`, `3.7` and `3.8` and
`pytest` version 5. Python 2 is not supported.

`pytest-docker` was originally created by André Caron.

# Installation
Install `pytest-docker` with `pip` or add it to your test requirements. It is
recommended to install `docker-compose` python package directly in your
environment to ensure that it is available during tests. This will prevent
potential dependency conflicts that can occur when the system wide
`docker-compose` is used in tests.


# Usage
Here is an example of a test that depends on a HTTP service.

With a `docker-compose.yml` file like this (using the
[httpbin](https://httpbin.org/) service):

```yaml
version: '2'
services:
  httpbin:
    image: "kennethreitz/httpbin"
    ports:
      - "8000:80"
```

You can write a test like this:

```python
import pytest
import requests

from requests.exceptions import ConnectionError


def is_responsive(url):
    try:
        response = requests.get(url)
        if response.status_code == 200:
            return True
    except ConnectionError:
        return False


@pytest.fixture(scope="session")
def http_service(docker_ip, docker_services):
    """Ensure that HTTP service is up and responsive."""

    # `port_for` takes a container port and returns the corresponding host port
    port = docker_services.port_for("httpbin", 80)
    url = "http://{}:{}".format(docker_ip, port)
    docker_services.wait_until_responsive(
        timeout=30.0, pause=0.1, check=lambda: is_responsive(url)
    )
    return url


def test_status_code(http_service):
    status = 418
    response = requests.get(http_service + "/status/{}".format(status))

    assert response.status_code == status
```

By default this plugin will try to open `docker-compose.yml` in your
`tests` directory. If you need to use a custom location, override the
`docker_compose_file` fixture inside your `conftest.py` file:

```python
import os
import pytest


@pytest.fixture(scope="session")
def docker_compose_file(pytestconfig):
    return os.path.join(str(pytestconfig.rootdir), "mycustomdir", "docker-compose.yml")
```

## Available fixtures
All fixtures have `session` scope.

### `docker_ip`

Determine the IP address for TCP connections to Docker containers.

### `docker_compose_file`

Get an absolute path to the  `docker-compose.yml` file. Override this fixture in
your tests if you need a custom location.

### `docker_compose_project_name`

Generate a project name using the current process PID. Override this fixture in
your tests if you need a particular project name.

### `docker_services`

Start all services from the docker compose file (`docker-compose up`).
After test are finished, shutdown all services (`docker-compose down`).

## Options

Optional features can be enabled/configured with environment variables:

### `PYTEST_DOCKER_LOG_DIR`

If given the complete logs of the Docker Compose application under test will be exported there.

### `PYTEST_DOCKER_HOST`

If `pytest-docker` tests are run in containers themselves, the endpoints where the Compose application services under test may be reachable may differ. The `docker_services` fixture provides a method `endpoint_for` which will return the reachable endpoint (host and port) depending on the setting of `PYTEST_DOCKER_HOST`:

  * unset: host is `127.0.0.1` (localhost) or `DOCKER_HOST` if set and the ports are the exposed ports.
  * `_internal`: tests are expected to be run in the same bridged network as the Compose application under test. Thus the services are to be reached by their internal Docker host names (Compose service names) and their internal container ports.
  * Any other value is expected to be a reachable hostname or IP. This may be the case if the tests are run in a container on the `host` network. The ports are the exposed ports.


# Contributing
This pytest plug-in and its source code are made available to you under a MIT
license. It is safe to use in commercial and closed-source applications. Read
the license for details!

Found a bug? Think a new feature would make this plug-in more practical? We
welcome issues and pull requests!

When creating a pull request, be sure to follow this projects conventions - use
[black](https://github.com/psf/black) with default settings for formatting, run
tests with `python setup.py test` and make sure that all checks are passing.


CHANGELOG.md
# Changelog

## Version 0.8
- export container logs to directory specified by `PYTEST_DOCKER_LOG_DIR`
- configure where Compose services under test may be reached by setting `PYTEST_DOCKER_HOST`
- fixture for monkeypatching a whole session: `monkeypatch_session`

## Version 0.7.2
- Changes:
	- Update package README

## Version 0.7.1
- Minor packaging related fixes
	- Fix description content type in `setup.cfg`
	- Add `skip_cleanup` option to travis-ci settings

## Version 0.7.0
- Removed the fallback method that allowed running without Docker
- Package cleanup

## Version 0.6.0
- Added ability to return list of files from `docker_compose_file` fixture

## Version 0.3.0
- Added `--build` option to `docker-compose up` command to automatically
  rebuild local containers


tests/test_integration.py
import shutil
import subprocess
from os import path, getenv

import requests
from requests.exceptions import ConnectionError


def is_responsive(url):
    """Check if something responds to ``url``."""
    try:
        response = requests.get(url)
        if response.status_code == 204:
            return True
    except ConnectionError:
        return False


def test_main_fixtures_work(docker_ip, docker_services):
    """Showcase the power of our Docker fixtures!"""

    # Build URL to service listening on random port.
    url = "http://%s:%d/" % (docker_ip, docker_services.port_for("hello", 80))

    assert not getenv('DOCKER_HOST')
    assert not getenv('PYTEST_DOCKER_HOST')
    endpoint_host, endpoint_port = docker_services.endpoint_for("hello", 80)
    assert endpoint_host == '127.0.0.1'
    assert endpoint_port > 80

    docker_services.wait_until_responsive(
        check=lambda: is_responsive(url), timeout=30.0, pause=0.1
    )

    # Contact the service.
    response = requests.get(url)
    # this is set up in the test image
    assert response.status_code == 204


def test_containers_and_volumes_get_cleaned_up(
    testdir, tmpdir, docker_compose_file
):
    _copy_compose_files_to_testdir(testdir, docker_compose_file)

    project_name_file_path = path.join(str(tmpdir), "project_name.txt")
    testdir.makepyfile(
        """
    import subprocess

    def _check_volume_exists(project_name):
        check_proc = subprocess.Popen(
            "docker volume ls".split(),
            stdout=subprocess.PIPE,
        )
        assert project_name.encode() in check_proc.stdout.read()

    def _check_container_exists(project_name):
        check_proc = subprocess.Popen(
            "docker ps".split(),
            stdout=subprocess.PIPE,
        )
        assert project_name.encode() in check_proc.stdout.read()

    def test_whatever(docker_services, docker_compose_project_name):
        _check_volume_exists(docker_compose_project_name)
        _check_container_exists(docker_compose_project_name)
        with open('{}', 'w') as project_name_file:
            project_name_file.write(docker_compose_project_name)
    """.format(
            str(project_name_file_path)
        )
    )

    result = testdir.runpytest()
    result.assert_outcomes(passed=1)

    with open(str(project_name_file_path), "rb") as project_name_file:
        compose_project_name = project_name_file.read().decode()
    _check_volume_is_gone(compose_project_name)
    _check_container_is_gone(compose_project_name)


def _copy_compose_files_to_testdir(testdir, compose_file_path):
    directory_for_compose_files = testdir.mkdir("tests")
    shutil.copy(compose_file_path, str(directory_for_compose_files))

    container_build_files_dir = path.realpath(
        path.join(compose_file_path, "../containers")
    )
    shutil.copytree(
        container_build_files_dir, str(directory_for_compose_files) + "/containers"
    )


def _check_volume_is_gone(project_name):
    check_proc = subprocess.Popen("docker volume ls".split(), stdout=subprocess.PIPE)
    assert project_name.encode() not in check_proc.stdout.read()


def _check_container_is_gone(project_name):
    check_proc = subprocess.Popen("docker ps".split(), stdout=subprocess.PIPE)
    assert project_name.encode() not in check_proc.stdout.read()


tests/test_docker_services.py
import subprocess
from unittest import mock

import pytest
from pytest_docker.plugin import DockerComposeExecutor, Services, get_docker_services


def test_docker_services():
    """Automatic teardown of all services."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [b"", b"0.0.0.0:32770", b""]
        check_output.returncode = 0

        assert check_output.call_count == 0

        # The fixture is a context-manager.
        with get_docker_services(
            "docker-compose.yml", docker_compose_project_name="pytest123"
        ) as services:
            assert isinstance(services, Services)

            assert check_output.call_count == 1

            # Can request port for services.
            port = services.port_for("abc", 123)
            assert port == 32770

            assert check_output.call_count == 2

            # 2nd request for same service should hit the cache.
            port = services.port_for("abc", 123)
            assert port == 32770

            assert check_output.call_count == 2

        assert check_output.call_count == 3

    # Both should have been called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" up --build -d',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" port abc 123',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" down -v --remove-orphans',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
    ]


def test_docker_services_unused_port():
    """Complain loudly when the requested port is not used by the service."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [b"", b"", b""]
        check_output.returncode = 0

        assert check_output.call_count == 0

        # The fixture is a context-manager.
        with get_docker_services(
            "docker-compose.yml", docker_compose_project_name="pytest123"
        ) as services:
            assert isinstance(services, Services)

            assert check_output.call_count == 1

            # Can request port for services.
            with pytest.raises(ValueError) as exc:
                print(services.port_for("abc", 123))
            assert str(exc.value) == (
                'Could not detect port for "%s:%d".' % ("abc", 123)
            )

            assert check_output.call_count == 2

        assert check_output.call_count == 3

    # Both should have been called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "up --build -d",
            shell=True,
            stderr=subprocess.STDOUT,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "port abc 123",
            shell=True,
            stderr=subprocess.STDOUT,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" down -v --remove-orphans',
            shell=True,
            stderr=subprocess.STDOUT,
        ),
    ]


def test_docker_services_failure():
    """Propagate failure to start service."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [
            subprocess.CalledProcessError(1, "the command", b"the output")
        ]
        check_output.returncode = 1

        # The fixture is a context-manager.
        with pytest.raises(Exception) as exc:
            with get_docker_services(
                "docker-compose.yml", docker_compose_project_name="pytest123"
            ):
                pass

        # Failure propagates with improved diagnoatics.
        assert str(exc.value) == (
            'Command {} returned {}: """{}""".'.format("the command", 1, "the output")
        )

        assert check_output.call_count == 1

    # Tear down code should not be called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "up --build -d",
            shell=True,
            stderr=subprocess.STDOUT,
        )
    ]


def test_wait_until_responsive_timeout():
    clock = mock.MagicMock()
    clock.side_effect = [0.0, 1.0, 2.0, 3.0]

    with mock.patch("time.sleep") as sleep:
        docker_compose = DockerComposeExecutor(
            compose_files="docker-compose.yml", compose_project_name="pytest123"
        )
        services = Services(docker_compose)
        with pytest.raises(Exception) as exc:
            print(
                services.wait_until_responsive(
                    check=lambda: False, timeout=3.0, pause=1.0, clock=clock
                )
            )
        assert sleep.call_args_list == [mock.call(1.0), mock.call(1.0), mock.call(1.0)]
    assert str(exc.value) == ("Timeout reached while waiting on service!")


tests/containerized/pytest_in_container.sh
#!/bin/bash
set -o nounset
#set -o errexit
#set -o xtrace

function get_script_dir(){
    SCRIPT_DIR="$( cd "$( dirname "$( readlink -f "${BASH_SOURCE[0]}" )" )" >/dev/null 2>&1 && pwd )"
    echo $SCRIPT_DIR
}
SCRIPT_DIR=$(get_script_dir)

# @todo: copy the src
docker build $SCRIPT_DIR/pytest_docker -t pytest_docker
mkdir -p results

DEFAULT_ARGS=(
    --volume /var/run/docker.sock:/var/run/docker.sock
    --volume $SCRIPT_DIR/../../src:/pytest_docker/src
    --volume $SCRIPT_DIR/tests:/pytest_docker/tests/
    --volume $SCRIPT_DIR/../../results:/results
    -e PYTEST_DOCKER_LOG_DIR=/results
    -e PYTHONPATH=/pytest_docker/src:/pytest_docker/tests
    --workdir /pytest_docker
    #--env-file $SCRIPT_DIR/debug.env
)

test_host_network() {
    docker run "${DEFAULT_ARGS[@]}" \
        --network host \
        pytest_docker \
        python3 -m pytest \
            -v \
            --log-level DEBUG \
            --log-file /results/containerized_host.pytest.log \
            /pytest_docker/tests/_test_integration_container.py
}

test_bridge_outside() {
    # run pytest in a container in a separate user-defined bridged network (not the same as the SUT)

    # cleanup:
    docker network rm pytest_docker_tests

    docker network create pytest_docker_tests
    docker run "${DEFAULT_ARGS[@]}" \
        -e PYTEST_DOCKER_HOST=ibms-dev.local \
        --add-host ibms-dev.local:192.168.79.20 \
        --network pytest_docker_tests \
        pytest_docker \
        python3 -m pytest \
            -v \
            --log-level DEBUG \
            --log-file /results/containerized_outside.pytest.log \
            /pytest_docker/tests/_test_integration_container.py

    # cleanup:
    docker network rm pytest_docker_tests
}

test_bridge_inside() {
    # run pytest in a container in a the same user-defined bridged network as the SUT

    docker network create pytest_docker_tests
    docker run "${DEFAULT_ARGS[@]}" \
        -e PYTEST_DOCKER_HOST=_internal \
        --network pytest_docker_tests \
        pytest_docker \
        python3 -m pytest \
            -v \
            --log-level DEBUG \
            --log-file /results/containerized_inside.pytest.log \
            /pytest_docker/tests/_test_integration_container.py

    # cleanup:
    docker network rm pytest_docker_tests
}

test_host_network
test_bridge_outside
test_bridge_inside


src/pytest_docker/plugin.py
import contextlib
import os
import re
import subprocess
import time
import timeit
import logging

import attr

import pytest


def execute(command, success_codes=(0,)):
    """Run a shell command."""
    try:
        output = subprocess.check_output(command, stderr=subprocess.STDOUT, shell=True)
        status = 0
    except subprocess.CalledProcessError as error:
        output = error.output or b""
        status = error.returncode
        command = error.cmd

    if status not in success_codes:
        raise Exception(
            'Command {} returned {}: """{}""".'.format(
                command, status, output.decode("utf-8")
            )
        )
    return output


def get_docker_ip():
    # When talking to the Docker daemon via a UNIX socket, route all TCP
    # traffic to docker containers via the TCP loopback interface.
    docker_host = os.environ.get("DOCKER_HOST", "").strip()
    if not docker_host:
        return "127.0.0.1"

    match = re.match(r"^tcp://(.+?):\d+$", docker_host)
    if not match:
        raise ValueError('Invalid value for DOCKER_HOST: "%s".' % (docker_host,))
    return match.group(1)


@pytest.fixture(scope="session")
def docker_ip():
    """Determine the IP address for TCP connections to Docker containers."""

    return get_docker_ip()


@attr.s(frozen=True)
class Services:

    _docker_compose = attr.ib()
    _services = attr.ib(init=False, default=attr.Factory(dict))

    def port_for(self, service, container_port):
        """Return the "host" port for `service` and `container_port`.

        E.g. If the service is defined like this:

            version: '2'
            services:
              httpbin:
                build: .
                ports:
                  - "8000:80"

        this method will return 8000 for container_port=80.
        """

        # Lookup in the cache.
        cache = self._services.get(service, {}).get(container_port, None)
        if cache is not None:
            return cache

        output = self._docker_compose.execute("port %s %d" % (service, container_port))
        endpoint = output.strip().decode("utf-8")
        if not endpoint:
            raise ValueError(
                'Could not detect port for "%s:%d".' % (service, container_port)
            )

        # Usually, the IP address here is 0.0.0.0, so we don't use it.
        match = int(endpoint.split(":", 1)[1])

        # Store it in cache in case we request it multiple times.
        self._services.setdefault(service, {})[container_port] = match

        return match

    def endpoint_for(self, service, container_port):
        pytest_docker_host = os.getenv('PYTEST_DOCKER_HOST')
        if not pytest_docker_host:
            # not set: default pytest-docker behavior: services reachable on DOCKER_HOST on exposed port
            host = get_docker_ip()
            port = self.port_for(service, container_port)
        elif pytest_docker_host == '_internal':
            # tests run in same docker network: services reachable on their internal hostnames and ports
            host = service
            port = container_port
        else:
            # host explicitly given: services reachable on that host and exposed port
            host = pytest_docker_host
            port = self.port_for(service, container_port)
        return (host, port)

    def wait_until_responsive(self, check, timeout, pause, clock=timeit.default_timer):
        """Wait until a service is responsive."""

        ref = clock()
        now = ref
        while (now - ref) < timeout:
            if check():
                return
            time.sleep(pause)
            now = clock()

        raise Exception("Timeout reached while waiting on service!")


def str_to_list(arg):
    if isinstance(arg, (list, tuple)):
        return arg
    return [arg]


@attr.s(frozen=True)
class DockerComposeExecutor:

    _compose_files = attr.ib(converter=str_to_list)
    _compose_project_name = attr.ib()

    def execute(self, subcommand):
        command = "docker-compose"
        for compose_file in self._compose_files:
            command += ' -f "{}"'.format(compose_file)
        command += ' -p "{}" {}'.format(self._compose_project_name, subcommand)
        return execute(command)


@pytest.fixture(scope="session")
def docker_compose_file(pytestconfig):
    """Get an absolute path to the  `docker-compose.yml` file. Override this
    fixture in your tests if you need a custom location."""

    return os.path.join(str(pytestconfig.rootdir), "tests", "docker-compose.yml")


@pytest.fixture(scope="session")
def docker_compose_project_name():
    """Generate a project name using the current process PID. Override this
    fixture in your tests if you need a particular project name."""

    return "pytest{}".format(os.getpid())


@contextlib.contextmanager
def get_docker_services(docker_compose_file, docker_compose_project_name):
    docker_compose = DockerComposeExecutor(
        docker_compose_file, docker_compose_project_name
    )

    try:
        # Spawn containers.
        docker_compose.execute("up --build -d")
    except Exception as exception:
        export_logs(docker_compose)
        raise exception

    # Let test(s) run.
    yield Services(docker_compose)

    export_logs(docker_compose)
    # Clean up.
    docker_compose.execute('down -v --remove-orphans')


@pytest.fixture(scope="session")
def docker_services(docker_compose_file, docker_compose_project_name):
    """Start all services from a docker compose file (`docker-compose up`).
    After test are finished, shutdown all services (`docker-compose down`)."""

    with get_docker_services(
        docker_compose_file, docker_compose_project_name
    ) as docker_service:
        yield docker_service


@pytest.fixture(scope='session')
def monkeypatch_session():
    '''
    @see: https://github.com/pytest-dev/pytest/issues/1872#issuecomment-375108891
    '''

    from _pytest.monkeypatch import MonkeyPatch
    monkey_patch = MonkeyPatch()
    yield monkey_patch
    monkey_patch.undo()


def export_logs(docker_compose):
    # https://github.com/AndreLouisCaron/pytest-docker/issues/13#issuecomment-345497583
    log_dir = os.getenv('PYTEST_DOCKER_LOG_DIR')
    if log_dir is not None:
        # date_format = "%Y%m%d_%H%M%S"
        log_filename = "{}.docker.log".format(docker_compose._compose_project_name)
        log_path = os.path.join(log_dir, log_filename)
        logging.info('Exporting docker-compose logs to %s ...', log_path)
        # @todo: (re-)create log file on first (session) call, afterwards append
        docker_compose.execute("logs --no-color > {}".format(log_path))

        # feed docker log to logging (will at least be feed to junit by pytest)
        # @fixme: prevent double log line headers
        # with open(log, 'r') as compose_log:
        #     for line in compose_log:
        #         logging.info(line)
